Fix offset + globalization issues in StringSegment

GrabYourPitchforks opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/39140.

I also experimented with adding nullable annotations to this type to get some additional error checking, but I ended up reverting that commit because it was out of scope of this work.

Marked draft because I haven't run any tests whatsoever over this code. It's just a demonstration of how I'm thinking of solving the problem.

GrabYourPitchforks committed:
Nullable-enable StringSegment

GrabYourPitchforks committed:
Fix glob issues in StringSegment

GrabYourPitchforks committed:
Remove nullable annotations from StringSegment

eerhardt commented on src/libraries/Microsoft.Extensions.Primitives/src/ThrowHelper.cs:
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
```
(nit) necessary?

eerhardt commented on src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:
```diff
@@ -664,6 +618,33 @@ public override string ToString()
+            {
+                case StringComparison.CurrentCulture:
+                case StringComparison.CurrentCultureIgnoreCase:
+                    compareOptions = (CompareOptions)((int)comparisonType & 1);
```
(nit) This is kind of tricky, and not really intuitive to a reader. Is it written this way for performance?

A more straight forward way of writing it would be:

```C#

            switch (comparisonType)
            {
                case StringComparison.CurrentCulture:
                    compareOptions = CompareOptions.None;
                    return CultureInfo.CurrentCulture.CompareInfo;

                case StringComparison.CurrentCultureIgnoreCase:
                    compareOptions = CompareOptions.IgnoreCase;
                    return CultureInfo.CurrentCulture.CompareInfo;
```

GrabYourPitchforks replied:
Yeah, it was roughly copied from the switch statements [in string](https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L654). But after thinking through this further, I think those methods were trying to minimize codegen size, while we don't have that exact consideration here. We should be able to get away with the readable thing rather than the smallest-codegen thing.

davidfowl commented:
Performance tests!! 😄

GrabYourPitchforks commented:
@davidfowl There are lots of perf wins we could get from this type if we were willing to change some of its behavior. I leave that to a different issue.

davidfowl commented:
Yea, I just don't want regressions, this is in a hot path

GrabYourPitchforks commented:
What methods specifically do you need optimized? The only one that is likely to regress is `Equals(StringSegment)`, because it's now bouncing through an indirection. But I can hand-tune that if needed.

Reiterating my earlier comment: if this is _really_ in a hot path, seriously consider allowing some behavioral changes here. For instance, the fact that there's special-casing all over the place to distinguish between `default(StringSegment)` and `StringSegment.Empty` leads to unnecessary complexity + slower perf. Allowing the two to be treated as equivalent would optimize things.

davidfowl commented:
I haven't looked deeply enough to tell if that change will have an impact m, I'm just saying these types are only used in ASP.NET Core so make sure you do some performance testing

bbartels commented on src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:
```diff
@@ -146,24 +147,16 @@ public bool HasValue
-                diff = a.Length - b.Length;
-            }
+            // Use CompareInfo.Compare instead of MemoryExtensions.CompareTo because we need
+            // to distinguish between null and empty StringSegment values.
```
Can't HasValue be used to discriminate between null and empty segments?

bbartels replied:
Unless there are scenarios which I haven't considered the following passes all the StringSegment tests: https://github.com/bbartels/runtime/blob/bc223014a90472898906ed9a379e09e35dc62b7b/src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs#L151

Not sure about perf between `CompareInfo.Compare` and `ReadOnlySpan<char>.CompareTo`, though I assume CompareInfo.Compare is probably a wrapper around CompareTo?

GrabYourPitchforks committed:
Merge remote-tracking branch 'origin/master' into stringsegment

GrabYourPitchforks committed:
Code cleanup, perf improvements, improve globalization tests

GrabYourPitchforks commented:
@maryamariyan @eerhardt this is now ready for review. I've added a bunch of unit tests to check various edge cases, especially with respect to globalization. I've also added extra argument checks around the _Equals_ and _Contains_ fast-paths to mirror the checks that `string.Equals` performs in those same scenarios. Finally, all unsafe or unsafe-equivalent code has been removed from this type.

eerhardt commented on src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:
```diff
@@ -576,54 +558,40 @@ public int LastIndexOf(char value)
 
-            fixed (char* p = Buffer)
+            int i;
+            for (i = span.Length - 1; (uint)i < (uint)span.Length; i--)
```
> (uint)i < (uint)span.Length

Is this micro optimization worth the mental exercise someone reading this code has to go through? Honestly, this might be the first time I've ever seen a loop written this way.

GrabYourPitchforks replied:
It's to elide the bounds check, but we could write it using standard `i >= 0` syntax if desired. Last I checked, the JIT still emitted the bounds check, which would make the "simple" way a perf regression compared to the existing code.

eerhardt replied:
I'm definitely not an expert here, so double-check my work to make sure I didn't do anything wrong. But the current JIT'd code looks worse to me.

https://sharplab.io/#v2:C4LghgzgtgNAJiA1AHwAICYCMBYAUKgZgAJVMA2E9IgYSIG88imTjSKAjAe04BsiBZTAApSABiJgAlEUbMGuZoqIA3MACciEAA5gAdkQC8EgHQBBCAGUduoZIDcspUQCWu4C4cKnRAGacNQs6GmtbGADIAproA5sAAFkQAtESYdkRCAK6uwJJBADzpWW6S2nrhUbFxac6JiZKOSvLeSs4+6QDGcerGAJIQAOpxzsARVmDtEUKlugDazgC6kvVezUxNq06oAOxEwGoZEZ4bTAC+DYpnK0rbvmA8EIcNl80NhCTkRFy8AugimOJSGRXdZOVQaabBMBmSzWWxHbzZDznZh+AJBIzTcoxeJJFLVIgAPiMomqtWWzRBq1aHS6al6AyGIzGEym1jmi3Jx0pxxIOz2B3hG2e3mFmx2PjuD0FREuJyAA

```C#
    public static bool M1(string a) 
    {
        var span = a.AsSpan();
        int i;
        for (i = span.Length - 1; (uint)i < (uint)span.Length; i--)
        {
            if (char.IsWhiteSpace(span[i]))
            {
                return true;
            }
        }
        return false;
    }
        
    public static bool M2(string a) 
    {
        var span = a.AsSpan();
        int i;
        for (i = span.Length - 1; i >= 0; i--)
        {
            if (char.IsWhiteSpace(span[i]))
            {
                return true;
            }
        }
        return false;
    }
```

The left side is M1 and the right is M2:
![image](https://user-images.githubusercontent.com/8291187/105530286-2f718b00-5cad-11eb-9525-c70b9e310074.png)

GrabYourPitchforks replied:
That's x86 debug output. I'm trying to get x64 release output but it'll take a few minutes.

eerhardt reviewed (approved):
Just one comment on code that I found surprising.

LGTM. Nice work on filling out the unit tests for this type.

GrabYourPitchforks committed:
Simplify TrimEnd logic

GrabYourPitchforks commented:
@bbartels @eerhardt thanks both for your feedback! I've incorporated both of your suggestions and will let this auto-merge.

bbartels commented on src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:
```diff
@@ -146,24 +146,20 @@ public bool HasValue
+            else
+            {
+                CheckStringComparison(comparisonType); // must arg check before returning
+                return !a.HasValue ? (b.HasValue ? -1 : 0) : 1; // null sorts less than non-null, and two nulls sort as equal
```
Not suggesting it here, but a question in general, does the runtime guarantee the value of a boolean to be 1 if it is true (unless one themselves deliberately attempts to unsafe cast a given byte to a bool?

Because in such case couldn't one do `return Unsafe.As<bool, byte>(ref a.HasValue) - Unsafe.As<bool, byte>(ref b.HasValue);` to make the return value branchless?

GrabYourPitchforks replied:
Check the discussion over at https://github.com/dotnet/coreclr/pull/16138. Lots of good information regarding these kinds of optimizations over there.

bbartels replied:
Oh wow that is an amazing resource, thank you so much :)

GrabYourPitchforks committed:
skip some glob tests on netfx

GrabYourPitchforks commented:
I had forgotten that netfx (and netcore before 5.0) have some bugs w.r.t. the globalization _IsPrefix_ and _IsSuffix_ methods. I've suppressed on netfx the test cases which exercise those edge conditions.

GrabYourPitchforks commented:
Opened https://github.com/dotnet/runtime/issues/47374 to track CI failures. Continuing merge since CI failures are unrelated.

davidfowl commented:
@GrabYourPitchforks did you do any performance testing? 

@sebastienros Lets keep an eye out here for regressions in ASP.NET Core since we're the only consumers of this.

GrabYourPitchforks commented:
@davidfowl No formal benchmarking, but I looked at the codegen for hot methods like `Equals` and verified that things appear as expected. E2E testing will be more useful than microbenchmarking here.
